### PR TITLE
Vsnprintf

### PR DESCRIPTION
**Pull request by k-stachowiak**
## Description
This PR introduces publicly contributed improvement of handling the vsnprintf function in the library.
The function is only needed from the `debug.c` file, and the proper implementation would be selected at the compile time at the call site. This PR provides an `mbedtls_` symbol which delegates the implementation selection to the `platform.c` module like it would happen with the `printf`, `snprintf` etc.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
None

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Steps to test or reproduce
None

### CODE REVIEW

File: include/mbedtls/platform.h [hunk 1]
```diff
@@ -214,6 +217,41 @@ int mbedtls_platform_set_snprintf( int (*snprintf_func)( char * s, size_t n,
...
+ *   the destination buffer is too short.
+ */
+#if defined(_WIN32)
+/* For Windows (inc. MSYS2), we provide our own fixed implementation */
+int mbedtls_platform_win32_vsnprintf( char *s, size_t n, const char *fmt, va_list arg );
+#endif
+
+#if defined(MBEDTLS_PLATFORM_VSNPRINTF_ALT)
+#include <stdarg.h>
+extern int (*mbedtls_vsnprintf)( char * s, size_t n, const char * format, va_list arg );
```
Comment (mazimkhan) [inline include/mbedtls/platform.h]: Consider making a ```typedef``` for this function pointer.

Comment (k-stachowiak) [reply include/mbedtls/platform.h]: I see how this would help with the large number of arguments to the wrapped function, but I think it is better to maintain consistency with other platform functions, e.g. very similar `snprintf()`

File: library/platform.c [hunk 2]
```diff
@@ -128,6 +128,61 @@ int mbedtls_platform_set_snprintf( int (*snprintf_func)( char * s, size_t n,
 }
 #endif /* MBEDTLS_PLATFORM_SNPRINTF_ALT */
 
+#if defined(_WIN32)
+int mbedtls_platform_win32_vsnprintf( char *s, size_t n, const char *fmt, va_list arg )
```
Comment (mazimkhan) [inline library/platform.c]: This function can now be called from ```mbedtls_platform_win32_snprintf``` to avoid redundant code.

File: library/platform.c [hunk 3]
```diff
@@ -128,6 +128,61 @@ int mbedtls_platform_set_snprintf( int (*snprintf_func)( char * s, size_t n,
...
+    int ret;
+
+    /* Avoid calling the invalid parameter handler by checking ourselves */
+    if( s == NULL || n == 0 || fmt == NULL )
+        return( -1 );
+
+#if defined(_TRUNCATE)
+    ret = _vsnprintf_s( s, n, _TRUNCATE, fmt, arg );
+#else
+    ret = _vsnprintf( s, n, fmt, arg );
```
Comment (mazimkhan) [inline library/platform.c]: With or without truncation these functions can return ```-1``` when output buffer is not enough. On the other hand ```vsnprintf``` returns ```n``` or more. If the caller wants to use the return code for determining truncation then this function is not uniform with ```vsnprintf```. Consider making return code uniform. This is up for discussion @sbutcher-arm .

**Comment by yanesca** [comment]
The PR looks good to me in general. However, like @mazimkhan, I too would prefer to have a C99 compliant behaviour in our `mbedtls_platform_win32_vsnprintf()` implementation.

**Comment by sbutcher-arm** [comment]
@yanesca - Can you please approve the PR in the form? Thanks!

File: library/platform.c (hunk 3 above)
Comment (sbutcher-arm) [reply library/platform.c]: That's a fair point, but why on Windows are we calling `_vsnprintf()` and not `vsnprintf()` which is C99 compliant anyway?

To be clear - we need to define that `mbedtls_vsnprintf()` is C99 compliant, and our own implementations MUST conform to that. Unfortunately, this is a blocker on the PR.

Comment (mazimkhan) [reply library/platform.c]: [Here](https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l) it say:
```
Beginning with the UCRT in Visual Studio 2015 and Windows 10, vsnprintf is no longer identical to _vsnprintf. The vsnprintf function complies with the C99 standard; _vnsprintf is retained for backward compatibility with older Visual Studio code.
```
It might be recent than the [introduction](https://github.com/ARMmbed/mbedtls/commit/6c0c8e0d3de28e711811f4345464c112ad64d390) of ```_vsnprintf``` in mbedtls. May be time for change.

File: library/platform.c [hunk 4]
```diff
@@ -125,9 +125,9 @@ int mbedtls_platform_win32_vsnprintf( char *s, size_t n, const char *fmt, va_lis
         return( -1 );
 
 #if defined(_TRUNCATE)
-    ret = _vsnprintf_s( s, n, _TRUNCATE, fmt, arg );
+    ret = vsnprintf_s( s, n, _TRUNCATE, fmt, arg );
 #else
-    ret = _vsnprintf( s, n, fmt, arg );
+    ret = vsnprintf( s, n, fmt, arg );
```
Comment (mazimkhan) [inline library/platform.c]: Purpose of ```mbedtls_platform_win32_vsnprintf``` was to provide C99 complaint function. Now if we have ```vsnprintf``` on Windows then we don't require this function all together. 
Since @mpg originally added this function. I will ask for his opinion.

File: include/mbedtls/config.h [hunk 5]
```diff
@@ -2789,6 +2790,7 @@
 //#define MBEDTLS_PLATFORM_PRINTF_MACRO        printf /**< Default printf macro to use, can be undefined */
 /* Note: your snprintf must correclty zero-terminate the buffer! */
 //#define MBEDTLS_PLATFORM_SNPRINTF_MACRO    snprintf /**< Default snprintf macro to use, can be undefined */
+//#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf /**< Default snprintf macro to use, can be undefined */
```
Comment (mpg) [inline include/mbedtls/config.h]: Typo: snprintf -> vsnprintf :)

File: library/platform.c (hunk 4 above)
Comment (mazimkhan) [reply library/platform.c]: Do you mean to say that the statement [here](https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l) is not correct. Snippet:
```
Beginning with the UCRT in Visual Studio 2015 and Windows 10, vsnprintf is no longer identical to _vsnprintf. The vsnprintf function complies with the C99 standard; _vnsprintf is retained for backward compatibility with older Visual Studio code.
```
If we are not sure about Windows behaviour recently. We can write some tests to confirm C99 compliance. We also need to maintain uniform behaviour across new and old versions of Windows and VS.

Comment (mpg) [reply library/platform.c]: No, I'm not saying this statement is incorrect. I guess the important part is "last time I checked" which apparently was prior to Windows 10. That's excellent news, I wasn't aware of this, good that MS improved their C99 support in that area, thanks for letting me know!

Unfortunately, unless we explicitly support only VS2015+Windows10 and above, I'm afraid we can't fully enjoy the comfort of a C99-compliant vsnprint() just yet, can we?

I fully agree regarding tests. Currently we have `test_snprintf()` in `tests/suites/main_test.function`. We probably want something similar for `vsnprintf()` just to be sure.

Comment (mazimkhan) [reply library/platform.c]: So essentially we can change ```_WIN32``` to something that precisely tells when ```vsnprintf``` became C99 compliant on Windows.

Comment (mpg) [reply library/platform.c]: Yes, if we can detect that with an `#if` that would be great: if we have compliant versions of `vsnprintf` and `snprintf`, just use them directly, and otherwise, include the old code for building our own compliant versions and registering them as `mbedtls_(v)snprintf`.

Comment (sbutcher-arm) [reply library/platform.c]: So, in the `development` branch we're aiming to support Windows 7 and up, and in the LTS branches, `mbedtls-2.7` and `mbedtls-2.1` we're supporting the older versions of Windows.

We need to only invoke `vsnprintf()` if the version of Windows is correct. I'd suggest putting some guards around it using `_WIN32_WINNT` to check the version.

cc: @k-stachowiak

**Comment by k-stachowiak** [comment]
@mpg, @sbutcher-arm, @yanesca, @mazimkhan: I addressed the most recent review comments.

File: include/mbedtls/platform.h [hunk 6]
```diff
@@ -58,7 +58,7 @@ extern "C" {
 #endif
 #endif
 #if !defined(MBEDTLS_PLATFORM_STD_VSNPRINTF)
-#if defined(_WIN32)
+#if _WIN32_WINNT < 0x0A00 /* < Windows 10 */
```
Comment (mpg) [inline include/mbedtls/platform.h]: This test is not correct, it also catches Unix platforms: undefined macros evaluate to `0` according to the standard (I also confirmed it experimentally by adding a `#warning` inside that block: it get picked on Linux). You should first check if the symbol is defined before checking its value.

Also, the previous test was `defined(_WIN32)` and this one moves to `_WIN32_WINNT`: are there any circumstances when one it defined but not the other? If so, the test should for that.

**Comment by mpg** [comment]
Note: the IAR build in the CI failed, please investigate and fix.

**Comment by k-stachowiak** [comment]
@mpg The preprocessor checks were actually incorrect. I have fixed them to check for the _WIN32 being defined and the MSVC version being less than 2015 (a.k.a. 14.0, a.k.a. 1900)

File: include/mbedtls/platform.h [hunk 7]
```diff
@@ -58,7 +58,7 @@ extern "C" {
 #endif
 #endif
 #if !defined(MBEDTLS_PLATFORM_STD_VSNPRINTF)
-#if _WIN32_WINNT < 0x0A00 /* < Windows 10 */
+#if _WIN32 && _MSC_VER < 1900 /* Before MSVC 14.0 / Visual Studio 2015 */
```
Comment (mpg) [inline include/mbedtls/platform.h]: I'm not sure this is correct either. What if the platform is windows and but compiler is something other than MSVC? I think you need to check if `_MSC_VER` is defined before checking its value.

Also, to check if macro is defined, please use `defined(MACRO)` rather than just `MACRO`.

Finally, since this check is repeated a few times, perhaps we should defined a macro for it, eg `WIN32_WITH_SNPRINTF_FAMILY`

File: library/platform.c [hunk 8]
```diff
@@ -115,7 +115,7 @@ int mbedtls_platform_set_snprintf( int (*snprintf_func)( char * s, size_t n,
 }
 #endif /* MBEDTLS_PLATFORM_SNPRINTF_ALT */
 
-#if _WIN32_WINNT < 0x0A00 /* < Windows 10 */
+#if _WIN32 && _MSC_VER < 1900 /* Before MSVC 14.0 / Visual Studio 2015 */
```
Comment (mpg) [inline library/platform.c]: Why is the test on line 73 different from this one?

File: include/mbedtls/platform.h (hunk 7 above)
Comment (k-stachowiak) [reply include/mbedtls/platform.h]: Aggreed.

I can see that MSVC has got a conformance update in the Visual Studio 2015. It's claimed to fully support the C99 standard library except for features requiring missing compiler functionality. This makes it an outstanding release. I'd turn that condition into a compile time define pointing to the VS version rather than particular function family as the compliance changes have most probably been wider that just the niche printing functions.

File: library/platform.c (hunk 8 above)
Comment (k-stachowiak) [reply library/platform.c]: The reason is the change being outside of this PR's scope, however it has been discussed that the changes could be more general.

Unfortunately I've noticed that there is MSYS mentioned as the reason for just testing for _WIN32. Which makes me want to make sure we have got the right filter here, as it may turn out the precise test whether the custom implementation is necessary or not happens to be complicated beyond feasibility.

File: library/platform.c [hunk 9]
```diff
@@ -70,28 +70,15 @@ int mbedtls_platform_set_calloc_free( void * (*calloc_func)( size_t, size_t ),
 }
 #endif /* MBEDTLS_PLATFORM_MEMORY */
 
-#if defined(_WIN32)
+#if defined(MSVC_LESS_THAN_14)
```
Comment (sbutcher-arm) [inline library/platform.c]: I don't think this will work on MinGW. MinGW will define `_WIN32`but not MSVC.

However - I also don't think you need to know. If you keep `#if defined(_WIN32)`, there will still be a call to `mbedtls_vsnprintf()` and what happens there you can work out locally.

File: include/mbedtls/platform.h [hunk 10]
```diff
@@ -46,17 +46,32 @@ extern "C" {
  * \{
  */
 
+/* This macro tells if the current compiler is Visual Studio, before the 2015
+ * edition. It is an important piece of information due to the standard
+ * compliance issues breaking consistency across the supported platforms.
+ */
+#if defined(_MSC_VER) && _MSC_VER < 1900
+#define MSVC_LESS_THAN_14
```
Comment (sbutcher-arm) [inline include/mbedtls/platform.h]: Throughout, I don't find `MSVC_LESS_THAN_14` very readable, and prefer `#if defined(_MSC_VER) && _MSC_VER < 1900` which is more readable and obvious.

Comment (mpg) [reply include/mbedtls/platform.h]: Also, since this name is defined in a public header, it should be namespaced.

**Comment by k-stachowiak** [comment]
@mpg @sbutcher-arm Upon further inspection I think we should leave the test against just `_WIN32`. Mingw/MSYS2 compiled software will dynamically link to the common runtime, which doesn't depend on the compiler but on whatever is installed in the end user's OS.

**Comment by mpg** [comment]
@k-stachowiak I don't understand why it wouldn't be a good idea to append `&& ! (defined(_MSC_VER) && _MSC_VER >= 1900)` to the existing `defined(_WIN32)` tests. Indeed with MSYS/MinGW we don't know what runtime we'll be using so we should be conservative, but with MSVC we know, so why not take advantage of it?

Note: I think just `defined(_WIN32)` is correct in all cases, it's just suboptimal in a few cases, one of which is easy to detect.

**Comment by k-stachowiak** [comment]
@mpg It seems that the executables built with newer MSVC will link to newer common runtime by default and therefore it is safe to assume that the MSVC version determines the implementation of the `(v)snprintf` that will be used. Previously I was not certain about that, which is why I proposed a more general test. But I checked how it works and I agree that the additional test for the Microsoft's compiler version is beneficial.

**Comment by sbutcher-arm** [comment]
retest

**Comment by sbutcher-arm** [comment]
Failed in the CI on the occasionally failing timing test, so retesting!

**Comment by sbutcher-arm** [comment]
retest

**Comment by k-stachowiak** [comment]
@mazimkhan, @yanesca, @sbutcher-arm: could at least one of you (re-)review this PR after the recent changes, please?

**Comment by k-stachowiak** [comment]
@mazimkhan Unless it turns out I need to do a rebase, I'd rather not modify the history to avoid the necessity of re-reviewing the PR. In case of a rebase I'll gladly clean up the commit history.

**Comment by mazimkhan** [comment]
> @mazimkhan Unless it turns out I need to do a rebase, I'd rather not modify the history to avoid the necessity of re-reviewing the PR. In case of a rebase I'll gladly clean up the commit history.

Successive commits within this PR have reworked the same compile time check. I don't see the benefit in preserving history of the approaches that were not accepted in the review. Fix-up will make a clean history after in the main repo. Review in this PR is over and fixup should not require a re-review.

File: include/mbedtls/platform.h [hunk 11]
```diff
@@ -46,17 +46,36 @@ extern "C" {
  * \{
  */
 
+/* The older Microsoft Windows common runtime provides non-conforming
+ * implementations of some standard library functions, including snprintf
+ * and vsnprintf. This affects MSVC and MinGW builds.
+ */
+#if _WIN32 && !(defined(_MSC_VER) && _MSC_VER >= 1900)
+#define MBEDTLS_PLATFORM_HAS_CONFORMING_SNPRINTF 0
```
[the remark this answers is not in the document]
Comment (k-stachowiak) [reply include/mbedtls/platform.h]: It can be reworked to only depend on one symbol. I'll do so.

**Comment by sbutcher-arm** [comment]
This PR has failed the CI Jenkins release job. See [job #548 here](https://jenkins-internal.mbed.com/blue/organizations/jenkins/mbedtls-release/detail/mbedtls-release/548/pipeline).

Results:
```

+---------+---------------+--------------+------------+--------------------+------------------+--------------------+----------------+-------------------+
| Version | Configuration | Architecture | Retargeted |   shipped build    | shipped selftest |    cmake build     | cmake selftest | cmake test suites |
+---------+---------------+--------------+------------+--------------------+------------------+--------------------+----------------+-------------------+
|   2010  |    Release    |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2010  |    Release    |    Win32     |    True    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2010  |     Debug     |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2010  |     Debug     |    Win32     |    True    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2013  |    Release    |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2013  |    Release    |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2013  |    Release    |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2013  |    Release    |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2013  |     Debug     |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2013  |     Debug     |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2013  |     Debug     |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2013  |     Debug     |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2015  |    Release    |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2015  |    Release    |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2015  |    Release    |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2015  |    Release    |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2015  |     Debug     |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2015  |     Debug     |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2015  |     Debug     |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2015  |     Debug     |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2017  |    Release    |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2017  |    Release    |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2017  |    Release    |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2017  |    Release    |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2017  |     Debug     |    Win32     |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2017  |     Debug     |    Win32     |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
|   2017  |     Debug     |     x64      |   False    |        Pass        |       Pass       |        Pass        |      Pass      |        Pass       |
|   2017  |     Debug     |     x64      |    True    | Pass with warnings |       Pass       | Pass with warnings |      Pass      |        Pass       |
+---------+---------------+--------------+------------+--------------------+------------------+--------------------+----------------+-------------------+
29 configurations tested, 29 successful
script returned exit code 2
```

**Comment by sbutcher-arm** [comment]
I have no idea why I highlighted the Visual Studio results. The error is in the mingw build, where the test suites are failing with the following error for each test suite:
```
...
test_suite_aes.cbc.exe ............................................ the snprintf implementation is broken
Use of uninitialized value $skipped in subtraction (-) at scripts/run-test-suites.pl line 69.
Use of uninitialized value $tests in subtraction (-) at scripts/run-test-suites.pl line 69.
FAIL
test_suite_aes.cfb.exe ............................................ the snprintf implementation is broken
Use of uninitialized value $skipped in subtraction (-) at scripts/run-test-suites.pl line 69.
Use of uninitialized value $tests in subtraction (-) at scripts/run-test-suites.pl line 69.
FAIL
...
```

**Comment by sbutcher-arm** [comment]
retest

File: library/platform.c [hunk 12]
```diff
@@ -131,6 +118,62 @@ int mbedtls_platform_set_snprintf( int (*snprintf_func)( char * s, size_t n,
...
+ * Make dummy function to prevent NULL pointer dereferences
+ */
+static int platform_vsnprintf_uninit( char * s, size_t n,
+                                     const char * format, va_list arg )
+{
+    ((void) s);
+    ((void) n);
+    ((void) format);
+    ((void) arg);
+    return( 0 );
```
Comment (mpg) [inline library/platform.c]: I think this should return `-1` as dummy implementations should always return an error code, which `0` isn't. For example, in `x509.c` we check for errors with `if( ret < 0 || (size_t) ret >= n )` which would treat `0` as success.

Comment (k-stachowiak) [reply library/platform.c]: I agree, I've fixed this.

**Comment by k-stachowiak** [comment]
@manuel Thank you, I addressed your remarks. Could you please review the change I made?

**Comment by k-stachowiak** [comment]
@sbutcher-arm I resolved the MinGW CI issue. Could you please re-review the PR?

File: ChangeLog [hunk 13]
```diff
@@ -22,6 +22,9 @@ Bugfix
 Changes
    * Change the shebang line in Perl scripts to look up perl in the PATH.
      Contributed by fbrosson.
+   * Provide Mbed TLS wrapper for vsnprintf used in debug.c. Based on
```
Comment (sbutcher-arm) [inline ChangeLog]: Can we rephrase this to be:
> Provide an abstraction of vsnprintf to allow alternative implementations for platforms that don't provide it. Based on contributions by Joris Aerts in and Nathaniel Wesley Filardo.

**Comment by sbutcher-arm** [comment]
@k-stachowiak - The PR looks fine apart from some trivial cosmetic issues. The only thing I really care about is the ChangeLog, so please fix that.

I've started a Windows CI job, which is build [#708](https://jenkins-internal.mbed.com/view/mbed-tls/job/mbedtls-release/708/). Assuming that passes, we're all good and ready to merge.

**Comment by sbutcher-arm** [comment]
This has failed `all.sh` test script with the following error:
```
******************************************************************
* test/build: declared and exported names 
* Fri Sep 21 21:28:23 UTC 2018
******************************************************************
Analysing source code...
1373 macros
     178 enum-consts
1122 identifiers
912 exported-symbols

Exported symbols declared in header: PASS
Names of actual-macros: FAIL
__USE_MINGW_ANSI_STDIO
Names of enum-consts: PASS
Names of identifiers: PASS
Likely typos: PASS

Overall: FAILED
script returned exit code 1
```

Obviously needs some rework, so relabelling as 'needs work'.

**Comment by mpg** [comment]
retest

**Comment by k-stachowiak** [comment]
retest

**Comment by k-stachowiak** [comment]
@mpg, @sbutcher-arm I introduced yet another change, which fixes an issue of one of the MinGW configurations. Could you please re-review?

**Comment by sbutcher-arm** [comment]
Jenkins CI job #872 passes, so promoting to 'ready for merge'.

**Comment by sbutcher-arm** [comment]
The builedbot CI fails on the PR with ' the snprintf implementation is broken' :
```
UpdateCTestConfiguration  from :C:/buildbot/mbedtls/any_-_polarssl_win32__mingw_make_/build/DartConfiguration.tcl
Parse Config file:C:/buildbot/mbedtls/any_-_polarssl_win32__mingw_make_/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :C:/buildbot/mbedtls/any_-_polarssl_win32__mingw_make_/build/DartConfiguration.tcl
Parse Config file:C:/buildbot/mbedtls/any_-_polarssl_win32__mingw_make_/build/DartConfiguration.tcl
Test project C:/buildbot/mbedtls/any_-_polarssl_win32__mingw_make_/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
      Start  1: aes.ecb-suite

1: Test command: C:\buildbot\mbedtls\any_-_polarssl_win32__mingw_make_\build\tests\test_suite_aes.ecb.exe "--verbose"
1: Test timeout computed to be: 10000000
1: the snprintf implementation is broken
 1/71 Test  #1: aes.ecb-suite ....................***Failed    0.02 sec
```

Demoting to 'needs: work'.

**Comment by k-stachowiak** [comment]
I have applied a fix added earlier by @mpg in 9db2887672 . I would like to see how it does on the platforms we have in the CI before asking for a re-review.

**Comment by k-stachowiak** [comment]
retest

**Comment by k-stachowiak** [comment]
@sbutcher-arm @mpg I have pushed a fix to an issue that I discovered. Could you please re-review? Due to limited control over my test environment, I cannot reproduce the issue raised by the build bot. If you know how to retest it in that CI, could you please perform the test? I only tested it in a synthetic environment and confirmed both the issue and the fix that way.

**Comment by mpg** [comment]
@k-stachowiak Can you please push this as a branch to our restricted repo? That's how buildbot is triggered. Once you've done that, please ping me and I'll also trigger the relevant portion of or release job on CI which has more windows tests. Thanks!

**Comment by k-stachowiak** [comment]
@mpg, I triggered the buildbot test as you advised, and all the tests pass there, except for the freebsd builders.

**Comment by mpg** [comment]
@k-stachowiak Thanks! Don't worry about the freebsd buildbot slaves, they're dead and we have the same tests running on BSD on Jenkins.

I triggered a run of the release job on Jenkins in order to get the fullest windows coverage we can have. https://jenkins-internal.mbed.com/blue/organizations/jenkins/mbedtls-release/detail/mbedtls-release/924/pipeline Please check it when it's over (should be about two hours from now) and report the result here.

**Comment by mpg** [comment]
The release job on Jenkins has passed as well.

**Comment by k-stachowiak** [comment]
@sbutcher-arm I addressed the build issue, thanks for pointing it out!

**Comment by sbutcher-arm** [comment]
Approved, passing CI. Let's merge it this time!
